R/module_nested_tabs.R

```diff
@@ -178,56 +163,93 @@ srv_nested_tabs.teal_modules <- function(id, datasets, modules, is_module_specif
   })
 }
 
-#' @rdname module_nested_tabs
+#' @rdname module_teal_module
 #' @export
-srv_nested_tabs.teal_module <- function(id, datasets, modules, is_module_specific = TRUE,
+srv_teal_module.teal_module <- function(id,
+                                        data_rv,
+                                        datasets,
+                                        modules,
                                         reporter = teal.reporter::Reporter$new()) {
-  checkmate::assert_class(datasets, "FilteredData")
-  logger::log_trace("srv_nested_tabs.teal_module initializing the module: { deparse1(modules$label) }.")
-
+  logger::log_trace("srv_teal_module.teal_module initializing the module: { deparse1(modules$label) }.")
   moduleServer(id = id, module = function(input, output, session) {
-    if (!is.null(modules$datanames) && is_module_specific) {
-      datasets$srv_filter_panel("module_filter_panel")
+    active_datanames <- reactive({
+      req(data_rv())
+      if (is.null(modules$datanames) || identical(modules$datanames, "all")) {
+        teal_data_datanames(data_rv())
+      } else {
+        include_parent_datanames(
+          modules$datanames,
+          teal.data::join_keys(data_rv())
+        )
+      }
+    })
+    if (is.null(datasets)) {
+      datasets <- eventReactive(data_rv(), {
+        logger::log_trace("srv_teal_module@1 initializing module-specific FilteredData")
+
+        # Otherwise, FilteredData will be created in the modules' scope later
+        progress_data <- Progress$new(
+          max = length(unlist(module_labels(modules)))
+        )
+        on.exit(progress_data$close())
+        progress_data$set(message = "Preparing data filtering", detail = "0%")
+        filtered_data <- teal_data_to_filtered_data(data_rv(), datanames = active_datanames())
+        filtered_data
+      })
     }
 
+    # manage module filters on the module level
+    # important:
+    #   filter_manager_module_srv needs to be called before filter_panel_srv
+    #   Because available_teal_slices is used in FilteredData$srv_available_slices (via srv_filter_panel)
+    #   and if it is not set, then it won't be available in the srv_filter_panel
```

Comment:
it would be great to enforce this programatically

Reply:
What we can do is to add this into `srv_filter_panel(id, datasets, active_datanames`

```r
assert_class(isolate(datasets()$get_available_teal_slices()), "reactive")
```